File: 리트코드/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int longest;
    public int diameterOfBinaryTree(TreeNode root) {
        dfs(root);
        return this.longest;

    }
    public int dfs(TreeNode root){
        if(root == null){
            return -1;
        }
        
        int left = dfs(root.left);
        int right = dfs(root.right);

        this.longest = Math.max(this.longest, left + right + 2);
        return Math.max(left,right) + 1;

    }
}

File: 리트코드/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun diameterOfBinaryTree(root: TreeNode?): Int {
        var longest:Int = 0;
        fun dfs(node: TreeNode?): Int{
            if(node == null)
                return -1
            
            val left = dfs(node.left)
            val right = dfs(node.right)

            longest = Math.max(longest, left + right + 2)
            return Math.max(left,right) + 1
        }
        
        dfs(root);
        return longest
    }
}